UI: Upgrade manager to react 18

ndelangen opened the pull request:
Replaces: https://github.com/storybookjs/storybook/pull/21673

Closes #20378 
Closes #22217

What i did:

Upgrade react and some testing deps then fix the fallout.

## Checklist

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to make sure your PR is ready to be reviewed. -->

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->

agriffis committed:
chore: upgrade to react 18

This is just the first step. Lots of stuff borked

```
yarn up react @types/react @types/react-dom @testing-library/react
```

Immediate issues with mismatched react-dom that wasn't listed explicitly
in package.json before, so

```
yarn add react-dom
```

agriffis committed:
fix: call react 18 createRoot instead of ReactDOM.render

agriffis committed:
chore: update react-test-renderer

agriffis committed:
fix: kill enzyme test, not compatible with react 18

agriffis committed:
fix: update typing since react 18

agriffis committed:
fix: ignore ReactDOM.render warning for PreviewWeb.integration.test.ts

agriffis committed:
chore: update to @testing-library/user-event 14

agriffis committed:
fix: link tests async with user-event 14

agriffis committed:
fix: update ui exports

valentinpalkovic committed:
Merge remote-tracking branch 'origin/next' into chore/react-18

ndelangen committed:
Merge branch 'next' into chore/react-18

ndelangen committed:
complete the upgrade

ndelangen committed:
solve type change upstream

ndelangen committed:
add type

ndelangen committed:
fix type problem after upgrade

ndelangen committed:
type fixes

ndelangen committed:
fix type issues

ndelangen committed:
fixes for types

ndelangen committed:
fixes for types

ndelangen committed:
fixes for types

ndelangen committed:
type fixes

ndelangen committed:
fix types

ndelangen committed:
types fixes

kasperpeulen committed:
Update svelte check

kasperpeulen committed:
Fix react type test

ndelangen committed:
this package depends on React16 and can't be pre-bundled

ndelangen committed:
fix types after react18 upgrade

ndelangen committed:
fixing types

ndelangen committed:
fixes for types

valentinpalkovic committed:
Remove React 18 type overrides and add additional types to fix lib

valentinpalkovic committed:
Fix theme type export to ensure proper functionality of Storybook themes

ndelangen committed:
Merge branch 'next' into chore/react-18

ndelangen committed:
fix sandbox script after some upgrades broke it 🤬

ndelangen committed:
Merge branch 'next' into chore/react-18

ndelangen committed:
sync version of react-dom

valentinpalkovic committed:
Fix interaction tests on chromatic

ndelangen committed:
Merge branch 'next' into chore/react-18

ndelangen committed:
Merge branch 'next' into chore/react-18

valentinpalkovic committed:
Merge branch 'next' into chore/react-18

ndelangen committed:
Merge branch 'next' into chore/react-18

ndelangen committed:
upgrades

ndelangen committed:
fix typing issue

ndelangen committed:
fixes

ndelangen committed:
Merge branch 'next' into chore/react-18

ndelangen committed:
Merge branch 'next' into chore/react-18

ndelangen committed:
Merge branch 'next' into pr/agriffis/21673

ndelangen committed:
Merge branch 'release-8-0' into pr/agriffis/21673-2

ndelangen committed:
Merge branch 'norbert/remove-storyshots' into pr/agriffis/21673-2

ndelangen committed:
fixes

ndelangen committed:
add exceptions for ts

ndelangen committed:
change tests

ndelangen committed:
fix linting

ndelangen committed:
Merge branch 'norbert/remove-storyshots' into pr/agriffis/21673-2

ndelangen committed:
Merge branch 'norbert/remove-storyshots' into pr/agriffis/21673-2

ndelangen committed:
skip test, over commenting it out

ndelangen committed:
Merge branch 'release-8-0' into pr/ndelangen/24511

ndelangen committed:
Merge branch 'release-8-0' into norbert/react18

valentinpalkovic reviewed (approved):
Code-wise LGTM! Do you want me to test something in particular @ndelangen?

ndelangen committed:
fix check

cdedreuille reviewed (approved):
Thanks a lot! LGTM.

ndelangen committed:
ensure the preview is done loading after navigation

JReinhold commented on code/addons/interactions/src/typings.d.ts:
```diff
@@ -0,0 +1,3 @@
+declare namespace React {
+  type RefForwardingComponent<T = unknown, S = unknown> = any;
```
I don't see this being used anywhere explicitly?

JReinhold commented on code/lib/react-dom-shim/src/preset.ts:
```diff
@@ -8,15 +8,15 @@ export const webpackFinal = async (config: any, options: Options) => {
   const isReact18 = version.startsWith('18') || version.startsWith('0.0.0');
   const useReact17 = legacyRootApi ?? !isReact18;
-  if (useReact17) return config;
+  if (!useReact17) return config;
 
   return {
     ...config,
     resolve: {
       ...config.resolve,
       alias: {
         ...config.resolve?.alias,
-        '@storybook/react-dom-shim': '@storybook/react-dom-shim/dist/react-18',
+        '@storybook/react-dom-shim': '@storybook/react-dom-shim/dist/react-16',
       },
     },
   };
```
something seems confusing to me in this file, `useReact17` but then shim with React 16? Can we clear that up somehow?

ndelangen replied:
Yes, great question, I do believe the "react17" part is very confusing, but that already existed.
I can explain what's going on:

This is about docs/preview.

The default is now react18.
You can see that by looking at the package.json of this `@storybook/react-dom-shim`.

So we do NOT have to alias when the user is using react18.
We do need to alias if the user is using react16 or 17

There's also a feature flag to force the legacyRootApi (which does no longer exist in react18, so that only applies to react17, because react16 can only do the legacyRoot thing.

JReinhold reviewed (approved):
A few questions, LGTM! 💪

ndelangen committed:
fix state & e2e tests

The state merging of useAddonState needed a fix because the merger function wasn't called with the correct data, due to react18 using async rendering, making this.state be the old state.

ndelangen committed:
simplify

ndelangen committed:
fix tests

ndelangen committed:
delete redundant typings file

ndelangen committed:
Merge branch 'release-8-0' into norbert/react18

ndelangen committed:
fix bench script

ndelangen committed:
Merge branch 'norbert/react18' of github.com:storybookjs/storybook into norbert/react18
